### PR TITLE
Refactor Transition to become a protocol

### PR DESCRIPTION
**Pull request by kylef**
This pull request brings substantial **breaking** changes to the current API of the Representor. It turns Transition into a Swift protocol. This means you have to create a concrete implementation of a transition for each transition protocol used. The representor now becomes a generic structure with it's defined Transition.

I have implemented a concrete HTTP transition in this pull request, and anyone using other protocols can build their own concrete implementation of their protocol. A transition has a transition builder implementation too (so it can introduce builder methods at the transition level, such as setting a HTTP method in the case of HTTPTransition.

For example, to create a HTTP based Representor:

``` swift
Representor<HTTPTransition> { builder in
  builder.addTransition("create", "/users") { builder in
    builder.method = "POST"
    builder.contentType = "application/json"
    builder.addAttribute("username")
    builder.addAttribute("first_name")
    builder.addAttribute("last_name")
  }
}
```

There are also breaking changes to bring the HTTP deserialiser out of the Representor itself, since the Representor is now a generic type. It isn't possible to store static variables on generic types, therefore I had to introduce a new structure called `HTTPDeserialization` which exposes `deserializers` and `preferredContentTypes` to allow the user to customise and add their own deserialisers for HTTP.

In previous versions, it was possible to import just the "Core" structures when using the representor if you didn't require or want the Builder DSL that I created. Unfortunately this is no longer possible and the builder has been moved from an extension to part of the core structures. This is purely due to a bug in the Swift compiler, when mangling method symbols of generic types in structure extensions (took me a while to figure out this bug was only exposed in extensions).

@zdne please review and merge.

### CODE REVIEW

**Comment by kylef** [comment]
@fosrias @smizell I think we should introduce a section on HTTP transitions to the charter. Defining the types of HTTP specific elements. This pull-request has introduced these two elements:
- method (the HTTP Method that should be used to make the request)
- contentType (the suggested contentType that should be used to make the request)

@smizell The Python Siren adapter should work the same way.

File: Representor/HTTP/HTTPTransition.swift [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+/** An implementation of the Transition protocol for HTTP. */
+public struct HTTPTransition : Transition {
+    public typealias Builder = HTTPTransitionBuilder
+
+    public let uri:String
+
+    /// The HTTP Method that should be used to make the request
+    public let method:String = "POST"
+    /// The suggested contentType that should be used to make the request
+    public let contentType:String?
```
Comment (kylef) [inline Representor/HTTP/HTTPTransition.swift]: After thinking about this some more, maybe we should extend on Siren's understanding of this and it should become an array so we can suggest multiple types in order of preference.

``` swift
/// An array of suggested content types to use when making a request. Ordered by preference.
public let suggestedContentTypes:[String]
```

Opinions?

File: Representor/Transition.swift [hunk 2]
```diff
@@ -30,27 +30,14 @@ public func ==<T : AnyObject>(lhs:InputProperty<T>, rhs:InputProperty<T>) -> Boo
 public typealias InputProperties = Dictionary<String, InputProperty<AnyObject>>
 
 /** Transition instances encapsulate information about interacting with links and forms. */
-public struct Transition : Equatable, Hashable {
-  public let uri:String
+public protocol Transition : Equatable, Hashable {
```
Comment (robrix) [inline Representor/Transition.swift]: Swift convention would be to name the protocol `TransitionType`.

Comment (kylef) [inline Representor/Transition.swift]: Thanks for picking up on that, I certainly did this the wrong way round `<TransitionType : Transition>`.

**Comment by smizell** [comment]
> The Python Siren adapter should work the same way.

@kylef I think it might be helpful to go off what [Verbose](http://verbose.readthedocs.org/en/latest/specification.html) has, as I've already worked through some of these differences. It has an array for request and response types. It also has arrays for several other things, even `rels`.

File: Representor/HTTP/HTTPTransition.swift (hunk 1 above)
Comment (zdne) [inline Representor/HTTP/HTTPTransition.swift]: I agree (+1 for array). What you think @smizell @fosrias ?

**Comment by zdne** [comment]
@kylef per @smizell  comment – do you want to revisit this or is it good to merge?

File: Representor/HTTP/HTTPTransition.swift (hunk 1 above)
Comment (fosrias) [inline Representor/HTTP/HTTPTransition.swift]: Array makes sense to me.

Comment (smizell) [inline Representor/HTTP/HTTPTransition.swift]: :+1: array

**Comment by kylef** [comment]
@zdne Added suggested content types as an array and I think this is good to merge.

**Comment by zdne** [comment]
~~@kylef  please add link to the Trello internal tracking card to every PR (where applicable)~~

**Comment by fosrias** [comment]
@zdne Do we want to add internal trello cards to OSS projects?

**Comment by zdne** [comment]
@fosrias per our offline discussion lets not do that (internal tracking cards on OSS project) I will communicate this to @kylef and @pksunkara tomorrow

**Pull request merged**